checker/src/main/java/org/checkerframework/checker/regex/RegexTransfer.java

```diff
@@ -246,4 +200,75 @@ public RegexTransfer(CFAbstractAnalysis<CFValue, CFStore, CFTransfer> analysis)
+            MethodInvocationNode n,
+            TransferResult<CFValue, CFStore> result,
+            RegexAnnotatedTypeFactory factory) {
+        CFStore thenStore = result.getRegularStore();
+        CFStore elseStore = thenStore.copy();
+        ConditionalTransferResult<CFValue, CFStore> newResult =
+                new ConditionalTransferResult<>(result.getResultValue(), thenStore, elseStore);
+        JavaExpression firstParam = JavaExpression.fromNode(n.getArgument(0));
+
+        if (thenStore.getValue(firstParam) != null) {
+            thenStore.clearValue(firstParam);
+        }
+        AnnotationMirror regexNNGroupsAnnotation = getRegexNNGroupsAnno(factory, n.getArguments());
+        thenStore.insertValue(firstParam, regexNNGroupsAnnotation);
+        return newResult;
+    }
+
+    /**
+     * The new store with modified on the basis of {@code asRegex} refinement.
+     *
+     * @param n the {@code RegexUtil.isRegex} method invocation
+     * @param result the store just before the invocation is encountered
+     * @param factory the type factory to create relevant types
+     * @return the new store after refining the type of the string that was refined.
+     */
+    private TransferResult<CFValue, CFStore> getAsRegexModifiedStore(
+            MethodInvocationNode n,
+            TransferResult<CFValue, CFStore> result,
+            RegexAnnotatedTypeFactory factory) {
+        AnnotationMirror regexNNGroupsAnnotation = getRegexNNGroupsAnno(factory, n.getArguments());
+        CFValue newResultValue =
+                analysis.createSingleAnnotationValue(
+                        regexNNGroupsAnnotation, result.getResultValue().getUnderlyingType());
+        return new RegularTransferResult<>(newResultValue, result.getRegularStore());
+    }
+
+    /**
+     * Returns the {@code @RegexNNGroups} annotation with relevant field values.
+     *
+     * @param factory the type factory to create types
+     * @param nodes arguments passed to the {@code RegexUtil} functions
+     * @return the {@code @RegexNNGroups} annotation
+     */
+    private AnnotationMirror getRegexNNGroupsAnno(
+            RegexAnnotatedTypeFactory factory, List<Node> nodes) {
+        int groupCount = 0;
+        List<Integer> nonNullGroups = new ArrayList<>();
+        for (Node e : nodes) {
+            if (e instanceof IntegerLiteralNode) {
+                groupCount = ((IntegerLiteralNode) e).getValue();
+            } else if (e instanceof ArrayCreationNode) {
+                List<Node> nonNullNodes = ((ArrayCreationNode) e).getInitializers();
+                for (Node inits : nonNullNodes) {
+                    if (inits instanceof IntegerLiteralNode) {
+                        nonNullGroups.add(((IntegerLiteralNode) inits).getValue());
+                    }
+                }
+            }
+        }
+        return factory.createRegexNNGroupsAnnotation(groupCount, nonNullGroups);
```

Comment:
If there are no groups, this might as well create `@Regex` rather than `@RegexNNGroups`.